### PR TITLE
feat(tracing): add explicit support for OTEL env vars

### PR DESCRIPTION
**Pull request by BradleyChatha**
As we've seen the OTEL SDK doesn't seem to ever create traces when our application is deployed to GKE (unless we explicitly specify a sampler in the code).

The docs suggest it should be respecting the `OTEL_TRACES_SAMPLER` env var, however I tested this before my week off to no real results. Only reverting to an older verison of GoKit (before we removed the explicit AlwaysSample) seemed to have any change.

I'm not really sure why this issue is happening only in GKE.

So this PR simply implements manual handling of some of the tracing env vars instead of relying on the SDK to do this for us.

### CODE REVIEW

**Comment by BradleyChatha** [comment]
Considering this is an internal function, and the set of OTEL env vars likely won't change very much, I wonder if it's worth bothering with the `envGetter` interface (Which only exists so it can be mocked) and instead just use `(sampler string, samplerArg string)` as the params.

**Comment by andypwarren** [comment]
> Considering this is an internal function, and the set of OTEL env vars likely won't change very much, I wonder if it's worth bothering with the `envGetter` interface (Which only exists so it can be mocked) and instead just use `(sampler string, samplerArg string)` as the params.
> 
> This'd mean that `SetupGlobalState` would need to fetch the env vars though, don't know if that feels messy or not.

yeh the getter interface feels like a bit of an overkill. It might be simpler just to set the environment variables directly in the test case's. It does make the testing more complex but it would give an actual indication of it's use

**Comment by BradleyChatha** [comment]
I'll give that a go. It _shouldn't_ cause any issues with other tests either since `SetupGlobalState` tests need to make use of the `providerMutex`, so as long as the test sets it back to normal afterwards we shouldn't need to worry about test order or parallel tests.

File: tracing/otel/provider.go [hunk 1]
```diff
@@ -251,3 +264,38 @@ func (o *OtelProvider) tracerFromContext(ctx context.Context) trace.Tracer {
...
 	LogFromCtx(ctx context.Context) *zerolog.Logger
 }
+
+// Despite OTEL SDK's documentation, it doesn't seem to want to use the standard
+// OTEL env vars when deployed into GKE. Instead of trying to debug/find
+// the obscure piece of documentation on why this happens, we've instead
+// decided to handle some of the env vars ourself.
+func samplerFromEnv() (sdktrace.Sampler, error) {
+	sampler, _ := os.LookupEnv(otelSamplerEnvVar)
+	samplerArg, _ := os.LookupEnv(otelSamplerArgEnvVar)
```
Comment (andypwarren) [inline tracing/otel/provider.go:274]: If you don't need to distinguish between an empty value and an unset value than you can use `os.Getenv`

Comment (BradleyChatha) [reply tracing/otel/provider.go]: :D Knew there was one like that, but intelisense failed to show it (and I was too lazy to google).

File: tracing/otel/provider_test.go [hunk 2]
```diff
@@ -218,6 +218,70 @@ func TestStart(t *testing.T) {
...
 
+func TestSamplerFromEnv(t *testing.T) {
+	providerMutex.Lock()
+	defer func() {
+		os.Unsetenv(otelSamplerEnvVar)
+		os.Unsetenv(otelSamplerArgEnvVar)
+		providerMutex.Unlock()
+	}()
+
+	cases := map[string]struct {
```
Comment (andypwarren) [inline tracing/otel/provider_test.go:229]: Could you add a test case for the float parsing error

**Review by andypwarren — APPROVED** [review]

**Comment by BradleyChatha** [comment]
@andypwarren One last thing: Currently if the sampler env var exists, but is the wrong value, we just default to AlwaysSample.

I wonder if we should error on that case to prevent typos going undetected, and only default to AlwaysSample when the env var is empty (existing or otherwise).

**Comment by andypwarren** [comment]
> @andypwarren One last thing: Currently if the sampler env var exists, but is the wrong value, we just default to AlwaysSample.
> 
> I wonder if we should error on that case to prevent typos going undetected, and only default to AlwaysSample when the env var is empty (existing or otherwise).

yeh that's a good idea

**Pull request merged**